internal/prober.go

```diff
@@ -0,0 +1,101 @@
+package internal
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"sync/atomic"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+type probeType string
+
+const (
+	ready   probeType = "ready"
+	healthy probeType = "healthy"
+)
+
+// Prober represents health and readiness status of given component.
+type Prober struct {
+	logger log.Logger
+
+	ready   uint32
+	healthy uint32
+}
+
+// NewProber returns Prober representing readiness and healthiness of given component.
+func NewProber(logger log.Logger) *Prober {
+	return &Prober{logger: logger}
+}
+
+// HealthyHandler returns a HTTP Handler which responds health checks.
+func (p *Prober) HealthyHandler() http.HandlerFunc {
+	return p.probeHandler(p.isHealthy, healthy)
+}
+
+// ReadyHandler returns a HTTP Handler which responds readiness checks.
+func (p *Prober) ReadyHandler() http.HandlerFunc {
+	return p.probeHandler(p.isReady, ready)
+}
+
+func (p *Prober) probeHandler(probeFunc func() bool, t probeType) http.HandlerFunc {
+	return func(w http.ResponseWriter, _ *http.Request) {
+		if !probeFunc() {
+			http.Error(w, fmt.Sprintf("observatorium is NOT %v", t), http.StatusServiceUnavailable)
+			return
+		}
+		if _, err := io.WriteString(w, fmt.Sprintf("observatorium is %v", t)); err != nil {
+			level.Error(p.logger).Log("msg", "failed to write probe response", "probe type", t, "err", err)
+		}
+	}
+}
```

Comment:
Very nice!!!
One small comment, instead of the probeType, what's wrong with readinessHandler and livenessHandler?

Reply:
Thinking about this further while reviewing. I don't think this handler should be concerned with whether it is liveness or readiness. The handler endpoint is determined outside of this and then the real difference should be the funcs passed into this handler.

Reply:
Could you please elaborate this furthermore? What would you like to see as an API from the perspective of the consumer?

Reply:
Short nit: Do you think we can import this from thanos  `pkg/prober`? or is it too much overhead? We could then improve single version of the same probing thing.

Reply:
Essentially would be nice to have the improvements on Thanos side as well.

Reply:
I don't mind, I like the idea actually. Let's iterate on this one and then I can commit the same code to Thanos as well.

Reply:
The idea here is to have an arbitrary number of check functions being passed to that handler and once that handler is triggered via an HTTP request it will call these functions and report back whether or not those funcs returned successfully.
For that, there doesn't need to be anything specific in those handlers.

As an example, look at 
https://github.com/heptiolabs/healthcheck/blob/master/types.go#L22:6
https://github.com/heptiolabs/healthcheck/blob/master/handler.go#L75-L103

Reply:
@metalmatze Although I really like the idea to have multiple checkers for various checks, I think we should keep it simple for now, if we need more sophisticated checks we can just use `healthcheck` directly. That being said, I have refactored `prober` a bit more.